internal/util/clustering/clustering.go

```diff
@@ -0,0 +1,101 @@
+// Licensed to the LF AI & Data foundation under one
+// or more contributor license agreements. See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership. The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License. You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package clustering
+
+import (
+	"encoding/binary"
+	"math"
+	"strconv"
+
+	"github.com/cockroachdb/errors"
+
+	"github.com/milvus-io/milvus-proto/go-api/v2/commonpb"
+	"github.com/milvus-io/milvus-proto/go-api/v2/schemapb"
+	"github.com/milvus-io/milvus/internal/proto/internalpb"
+	"github.com/milvus-io/milvus/pkg/util/distance"
+	"github.com/milvus-io/milvus/pkg/util/funcutil"
+)
+
+const (
+	ClusteringCentroid    = "clustering.centroid"
+	ClusteringSize        = "clustering.size"
+	ClusteringId          = "clustering.id"
+	ClusteringOperationid = "clustering.operationID"
+
+	SearchClusteringFilterRatio = "clustering.filter_ratio"
+)
+
+func SearchClusteringOptions(kv []*commonpb.KeyValuePair) (*internalpb.SearchClusteringOptions, error) {
+	kvMap := funcutil.KeyValuePair2Map(kv)
+
+	clusteringOptions := &internalpb.SearchClusteringOptions{
+		FilterRatio: 0.5, // default
+	}
+
+	if clusterBasedFilterRatio, ok := kvMap[SearchClusteringFilterRatio]; ok {
+		b, err := strconv.ParseFloat(clusterBasedFilterRatio, 32)
```

Comment:
change to collection level config

Reply:
I suggest that user could set this config only if they are very sure about what they are doing, otherwise milvus would choose the ratio inside

Reply:
user config is convenient for performance benchmark. I'd like to make three config level. user client config > collection config > system config.